File: src/main/java/AoC2019_10.java
import static java.util.stream.Collectors.toMap;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.function.BinaryOperator;
import java.util.function.Function;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2019_10
        extends SolutionBase<List<AoC2019_10.Asteroid>, Integer, Integer> {
    
    private static final char ASTEROID = '#';
	
	private AoC2019_10(final boolean debug) {
		super(debug);
	}
	
	public static AoC2019_10 create() {
		return new AoC2019_10(false);
	}

	public static AoC2019_10 createDebug() {
		return new AoC2019_10(true);
	}

	@Override
    protected List<Asteroid> parseInput(final List<String> inputs) {
		final CharGrid grid = CharGrid.from(inputs);
        return grid.getAllEqualTo(ASTEROID)
            .map(asteroid -> Asteroid.from(asteroid, grid))
            .toList();
    }

    private Asteroid best(final List<Asteroid> asteroids) {
	    return asteroids.stream()
	        .sorted(Asteroid.byOthersCountDescending())
	        .findFirst().orElseThrow();
	}
	
	@Override
	public Integer solvePart1(final List<Asteroid> asteroids) {
        return best(asteroids).others().size();
	}
	
	@Override
	public Integer solvePart2(final List<Asteroid> asteroids) {
        return best(asteroids).others().values().stream()
	        .skip(199)
	        .limit(1)
	        .findFirst()
	        .map(n200 -> n200.getX() * 100 + n200.getY())
	        .orElseThrow();
	}

	@Samples({
	    @Sample(method = "part1", input = TEST1, expected = "8"),
	    @Sample(method = "part1", input = TEST2, expected = "33"),
	    @Sample(method = "part1", input = TEST3, expected = "35"),
	    @Sample(method = "part1", input = TEST4, expected = "41"),
	    @Sample(method = "part1", input = TEST5, expected = "210"),
	    @Sample(method = "part2", input = TEST5, expected = "802"),
	})
	public static void main(final String[] args) throws Exception {
	    AoC2019_10.create().run();
	}
	
	private static final String TEST1 = """
                .#..#
                .....
                #####
                ....#
                ...##
                """;
	private static final String TEST2 = """
                ......#.#.
                #..#.#....
                ..#######.
                .#.#.###..
                .#..#.....
                ..#....#.#
                #..#....#.
                .##.#..###
                ##...#..#.
                .#....####
                """;
	private static final String TEST3 = """
                #.#...#.#.
                .###....#.
                .#....#...
                ##.#.#.#.#
                ....#.#.#.
                .##..###.#
                ..#...##..
                ..##....##
                ......#...
                .####.###.
                """;
	private static final String TEST4 = """
                .#..#..###
                ####.###.#
                ....###.#.
                ..###.##.#
                ##.##.#.#.
                ....###..#
                ..#.#..#.#
                #..#.#.###
                .##...##.#
                .....#.#..
                """;
	private static final String TEST5 = """
                .#..##.###...#######
                ##.############..##.
                .#.######.########.#
                .###.#######.####.#.
                #####.##.#.##.###.##
                ..#####..#.#########
                ####################
                #.####....###.#.#.##
                ##.#################
                #####.##.###..####..
                ..######..##.#######
                ####.##.####...##..#
                .#####..#.######.###
                ##...#.##########...
                #.##########.#######
                .####.#.###.###.#.##
                ....##.##.###..#####
                .#.#.###########.###
                #.#.#.#####.####.###
                ###.##.####.##.#..##
                """;
	
	record Asteroid(Position position, Map<Double, Position> others) {
	    
	    public static Asteroid from(final Cell cell, final CharGrid grid) {
	        final Position asteroid = Asteroid.asPosition(cell);
	        return new Asteroid(asteroid, angles(grid, asteroid));
	    }
	    
	    public static Comparator<Asteroid> byOthersCountDescending() {
	        return (a, b) -> Integer.compare(b.others().size(), a.others().size());
	    }
	    
	    private static Position asPosition(final Cell cell) {
	        return Position.of(cell.getCol(), cell.getRow());
	    }
	    
	    private static int distanceSquared(final Position a, final Position b) {
	        final int dx = a.getX() - b.getX();
	        final int dy = a.getY() - b.getY();
	        return dx * dx + dy * dy;
	    }
	    
	    private static BinaryOperator<Position> closestTo(final Position pos) {
	        return (a, b) -> distanceSquared(a, pos) < distanceSquared(b, pos) ? a : b;
	    }
	    
	    private static double angle(final Position asteroid, final Position other) {
	        final double angle = Math.atan2(
	                other.getY() - asteroid.getY(),
	                other.getX() - asteroid.getX())
	            + Math.PI / 2;
	        return angle < 0 ? angle + 2 * Math.PI : angle;
	    }
	    
	    private static Map<Double, Position> angles(
	            final CharGrid grid, final Position asteroid
        ) {
	        return grid.getAllEqualTo(ASTEROID)
	            .map(Asteroid::asPosition)
	            .filter(other -> !other.equals(asteroid))
	            .collect(toMap(
	                    other -> angle(asteroid, other),
	                    Function.identity(),
	                    closestTo(asteroid),
	                    TreeMap::new));
	    }
	}
}

File: src/main/python/AoC2019_10.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 10
#

from __future__ import annotations

import math
import sys
from functools import reduce
from typing import Iterator
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import CharGrid

Position = tuple[int, int]
ASTEROID = "#"

TEST1 = """\
.#..#
.....
#####
....#
...##
"""
TEST2 = """\
......#.#.
#..#.#....
..#######.
.#.#.###..
.#..#.....
..#....#.#
#..#....#.
.##.#..###
##...#..#.
.#....####
"""
TEST3 = """\
#.#...#.#.
.###....#.
.#....#...
##.#.#.#.#
....#.#.#.
.##..###.#
..#...##..
..##....##
......#...
.####.###.
"""
TEST4 = """\
.#..#..###
####.###.#
....###.#.
..###.##.#
##.##.#.#.
....###..#
..#.#..#.#
#..#.#.###
.##...##.#
.....#.#..
"""
TEST5 = """\
.#..##.###...#######
##.############..##.
.#.######.########.#
.###.#######.####.#.
#####.##.#.##.###.##
..#####..#.#########
####################
#.####....###.#.#.##
##.#################
#####.##.###..####..
..######..##.#######
####.##.####...##..#
.#####..#.######.###
##...#.##########...
#.##########.#######
.####.#.###.###.#.##
....##.##.###..#####
.#.#.###########.###
#.#.#.#####.####.###
###.##.####.##.#..##
"""


def asteroids(grid: CharGrid) -> Iterator[Position]:
    return (
        (c, r)
        for r in range(grid.get_height())
        for c in range(grid.get_width())
        if grid.get_value_at(r, c) == ASTEROID
    )


class Asteroid(NamedTuple):
    position: Position
    others: dict[float, Position]

    @classmethod
    def from_(cls, position: Position, grid: CharGrid) -> Asteroid:
        return Asteroid(position, Asteroid.angles(grid, position))

    @classmethod
    def angles(
        cls, grid: CharGrid, asteroid: Position
    ) -> dict[float, Position]:
        def merge(
            d: dict[float, Position], other: Position
        ) -> dict[float, Position]:

            def angle(asteroid: Position, other: Position) -> float:
                angle = (
                    math.atan2(other[1] - asteroid[1], other[0] - asteroid[0])
                    + math.pi / 2
                )
                return angle + 2 * math.pi if angle < 0 else angle

            def distance_squared(a: Position, b: Position) -> int:
                dx = a[0] - b[0]
                dy = a[1] - b[1]
                return dx**2 + dy**2

            a = angle(asteroid, other)
            d[a] = (
                min(
                    other,
                    d[a],
                    key=lambda x: distance_squared(x, asteroid),
                )
                if a in d
                else other
            )
            return d

        return reduce(
            merge,
            filter(lambda x: x != asteroid, asteroids(grid)),
            dict[float, Position](),
        )


Input = list[Asteroid]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        grid = CharGrid.from_strings(list(input_data))
        return [Asteroid.from_(p, grid) for p in asteroids(grid)]

    def best(self, asteroids: list[Asteroid]) -> Asteroid:
        return max(asteroids, key=lambda a: len(a.others))

    def part_1(self, asteroids: Input) -> Output1:
        return len(self.best(asteroids).others)

    def part_2(self, asteroids: Input) -> Output2:
        d = self.best(asteroids).others
        x, y = d[sorted(d.keys())[199]]
        return x * 100 + y

    @aoc_samples(
        (
            ("part_1", TEST1, 8),
            ("part_1", TEST2, 33),
            ("part_1", TEST3, 35),
            ("part_1", TEST4, 41),
            ("part_1", TEST5, 210),
            ("part_2", TEST5, 802),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2019, 10)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
